crates/astria-celestia-client/src/blob_space.rs

```diff
@@ -0,0 +1,320 @@
+        public_key
+            .verify(signature, &data_bytes)
+            .map_err(SignedNamespaceDataVerification::Verification)?;
+        Ok(Self::from_unverified_unchecked(unverified))
+    }
+}
+
+impl<D: Serialize> TryFrom<UnverifiedSignedNamespaceData<D>> for SignedNamespaceData<D> {
+    type Error = SignedNamespaceDataVerification;
+
+    fn try_from(unverified: UnverifiedSignedNamespaceData<D>) -> Result<Self, Self::Error> {
+        Self::try_from_unverified(unverified)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SignedNamespaceDataVerification {
+    #[error("failed verifying the serialized data with the given verification key and signature")]
+    Verification(#[source] ed25519_consensus::Error),
+}
+
+/// A shadow of `SignedNamespaceData` whose data has not been verified against its
+/// contained pubic key and signature.
+///
+/// This types primary use is for serialization and deserialization.
+/// Use [`Self::try_into_verified`] to verify its data and obtain a [`SignedNamespaceData`].
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UnverifiedSignedNamespaceData<D> {
+    pub data: D,
+    #[serde(with = "crate::blob_space::_serde::verification_key")]
+    pub public_key: VerificationKey,
+    #[serde(with = "crate::blob_space::_serde::signature")]
+    pub signature: Signature,
+}
+
+impl<D> UnverifiedSignedNamespaceData<D> {
+    fn from_verified(verified: SignedNamespaceData<D>) -> Self {
+        let SignedNamespaceData {
+            data,
+            public_key,
+            signature,
+        } = verified;
+        Self {
+            data,
+            public_key,
+            signature,
+        }
+    }
+}
+
+impl<D: Serialize> UnverifiedSignedNamespaceData<D> {
+    pub fn try_into_verified(
+        self,
+    ) -> Result<SignedNamespaceData<D>, SignedNamespaceDataVerification> {
+        SignedNamespaceData::try_from_unverified(self)
+    }
+}
+
+impl<D> From<SignedNamespaceData<D>> for UnverifiedSignedNamespaceData<D> {
+    fn from(verified: SignedNamespaceData<D>) -> Self {
+        Self::from_verified(verified)
+    }
+}
+
+/// SequencerNamespaceData represents the data written to the "base"
+/// sequencer namespace. It contains all the other chain IDs (and thus, namespaces) that were
+/// also written to in the same block.
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct SequencerNamespaceData {
+    pub block_hash: Hash,
+    pub header: Header,
+    pub last_commit: Option<Commit>,
+    pub rollup_chain_ids: Vec<ChainId>,
+    pub action_tree_root: [u8; 32],
+    pub action_tree_root_inclusion_proof: InclusionProof,
+    pub chain_ids_commitment: [u8; 32],
+}
+
+/// RollupNamespaceData represents the data written to a rollup namespace.
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct RollupNamespaceData {
+    pub block_hash: Hash,
+    pub chain_id: ChainId,
+    pub rollup_txs: Vec<Vec<u8>>,
+    pub inclusion_proof: InclusionProof,
+}
```

Comment:
why are these here? they aren't celestia-specific, imo they should be in the types or proto crate